services/src/main/java/io/druid/cli/CliPeon.java

```diff
@@ -205,17 +207,28 @@ public void run()
   {
     try {
       Injector injector = makeInjector();
-
       try {
-        Lifecycle lifecycle = initLifecycle(injector);
-
+        final Lifecycle lifecycle = initLifecycle(injector);
+        Runtime.getRuntime().addShutdownHook(
+            new Thread(
+                new Runnable()
+                {
+                  @Override
+                  public void run()
+                  {
+                    log.info("Running shutdown hook");
+                    lifecycle.stop();
+                  }
+                }
+            )
+        );
         injector.getInstance(ExecutorLifecycle.class).join();
-        lifecycle.stop();
```

Comment:
@gianm are you saying we should have both the shutdown hook and the `lifecycle.stop()` after the lifecycle.join()?